### ISSUE TITLE
[thi.ng/atom] - Extending classes for Observablehq

### ISSUE THREAD

**Issue by den1k** [thi-ng/umbrella#64]
I extended umbrella/atom to work idiomatically with observable (e.g. `viewof` and reactive cells), result [here](https://beta.observablehq.com/@den1k/thi-ng-umbrella-scaffold/2) - this now works with deeply nested objects!

extended Atom code [here](https://beta.observablehq.com/@den1k/atom)

But I'm struggling to get cursors to work, I think because I can't pass my the extended Atom class in the Cursor impl: https://github.com/thi-ng/umbrella/blob/master/packages/atom/src/cursor.ts#L97

If successful, I'd love to do the same for rstream and derived libs!

**Comment by den1k** [comment]
It looks like extending reactive parts of umbrella w/ the EventTarget interface will do the trick: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget - reactiveness comes from `dispatchEvent` which needs to run whenever values are updated, e.g. for `update`, `reset`, `resetIn` in atoms and `next` in streams and subscriptions.

Can this be done without overriding/redefining existing methods on `Atom` and `Stream`?

**Comment by den1k** [comment]
Some progress with `Stream`'s but couldn't get `Subscription`'s to work

https://beta.observablehq.com/d/d18d32f9c8545c65

**Comment by den1k** [comment]
HNY! Fixed the [first notebook](https://beta.observablehq.com/@den1k/thi-ng-umbrella-scaffold/2)... bundle.run is not very reliable.

The error in the second notebook is intentional, since I got stuck at that point.

Re: `viewof` here's a notebook explaining it: https://beta.observablehq.com/@mbostock/views-are-mutable-values

You're right I could just use addWatch or subscribe/transform (and I tried that) but observable is all about _seeing_ values update in realtime. Currently I can declare an atom or stream but would need to deref it to see changes. Or  a separate cell with an add-watch, as well as one for every cursor and derived subscription. This gets very verbose over time.

By implementing [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) which I did [successfully for atoms](https://beta.observablehq.com/@den1k/thi-ng-umbrella-scaffold/2) I can use `viewof` to get the atom, and otherwise automatically get the deref'd value of it, all while seeing values update reactively in the UI.
